src/python/pants/backend/jvm/tasks/jar_publish.py

```diff
@@ -1042,6 +1042,12 @@ def create_source_jar(self, target, open_jar, version):
     # ends up creating 2 jars one scala and other java both including the java_sources.
 
     def abs_and_relative_sources(target):
+      sources = target.payload.get_field('sources')
```

Comment:
Yes. The third test-case under jar_publish_integration was failing. That particular test case was publishing a .jar file from a java_protobuf_library. To be clear, it was not publishing something that depended on a java_protobuf_library, it was publishing the output of a java_protobuf_library directly. The relative sources weren't being correctly calculated from the protobuf library; they were being relativized as things like '../../../../../.pants.d/gobbledy-gook/gen/protoc/org/example/Foo.java' instead of 'org/example/Foo.java', which was breaking the jar-tool.

It's quite possible that there's something wrong instead with my code-gen changes that were causing this, but comparing my code to the previous code, I have no idea what or where that might be. 

target.get_sources_relative_to_buildroot() was breaking, and looking at its code, it was effectively taking the original (relative-path) sources, converting them to absolute paths, then relativizing them again against a different (wrong for protoc) path. So I cut out the middle-man and just used the original relative paths that were stored in the payload.

I don't whether the actual bug here was in the code I changed in jar_publish, in target.get_sources_relative_to_buildroot(), or in my protoc changes.